[docs] Asset checks guide (#24035)
dagster-io/dagster: docs/docs-beta/docs/guides/quality-testing/asset-checks.md

This guide covers the most common use cases for testing assets with asset checks and taking action based on the result.

<details>
<summary>Prerequisites</summary>
- Familiarity with [Assets](/concepts/assets)
</details>

## Testing assets with a single asset check

The example below defines a single asset check on an asset that fails if the `order_id` column of the asset contains a null value.

In this example, the asset check will run after the asset has been materialized, to ensure the quality of its data.

<CodeExample filePath="guides/data-assets/quality-testing/asset-checks/single-asset-check.py" language="python" title="Asset with a single asset check" />

## Testing assets with multiple asset checks

In most cases, checking the data quality of an asset will require multiple checks.

The example below defines two asset checks on an asset using the `multi_asset_check` decorator.
- one that fails if the `order_id` column of the asset contains a null value.
- another one that fails if the `item_id` column of the asset contains a null value.

> In this example, both asset check will run in a single operation after the asset has been materialized.

Reviewer: When you say, "a single operation", does that map to an `@op` under the hood?

Author: Yes exactly, see below

<img width="1399" alt="Screenshot 2024-08-29 at 1 46 41 PM" src="https://github.com/user-attachments/assets/9961d8e6-df0f-43ef-a145-9527e277f2b8">

Reviewer: TIL - thank you.

> <CodeExample filePath="guides/data-assets/quality-testing/asset-checks/multiple-asset-checks.py" language="python" title="Asset with multiple asset checks" />

Reviewer: A natural next-step after `multi_asset_check` would be a factory method where you have a list of columns to show how you can produce the AssetCheckSpec list. But perhaps we leave that for another guide.

Author: Done in [244ce77](https://github.com/dagster-io/dagster/pull/24035/commits/244ce771f6ec3bf85d09712f573739eb8550ccb3)

## Blocking downstream assets

> If an asset check fails, it may be useful to block the materialization of downstream assets to prevent the problem from spreading. To enable this behavior, set the `blocking` argument to `True` in the `asset_check` decorator.

Reviewer: ```suggestion
By default, materialization of downstream assets will continue, even if a parent's asset check fails. To block the materialization of downstream assets, set the `blocking` argument to `True` in the `asset_check` decorator.
```

I think emphasizing that it doesn't block "by default" is good

Author: Done in [e241f07](https://github.com/dagster-io/dagster/pull/24035/commits/e241f0776dea8f3c50bbee02c23173b6819e155f)

In the example bellow, when the `orders_id_has_no_nulls` check fails, the `augmented_orders` asset isn't materialized.

<CodeExample filePath="guides/data-assets/quality-testing/asset-checks/block-downstream-with-asset-checks.py" language="python" title="Block downstream assets when asset check fails" />

## Executing asset checks on a schedule

> In some case, it may be useful to run the asset checks on a schedule, separately from the job materializing the assets.

Reviewer: Can you provide an example of why someone might want to do this?

Are we able to reference the scheduled check result in a downstream asset, or trigger an alert?

Author: Done in [53c865e](https://github.com/dagster-io/dagster/pull/24035/commits/53c865e8a9578b9b49a38dc8e9dddbc50b0de42c)

In the example below, two jobs are defined, one for the asset and another one for the asset check. Using these jobs, schedules are defined to materialize the asset and execute the asset check independently.

<CodeExample filePath="guides/data-assets/quality-testing/asset-checks/asset-checks-with-schedule.py" language="python" title="Schedule asset checks separately from their asset" />

> ## Next steps

Reviewer: I don't believe we've linked to the blog in other guides, but this would be a good candidate to link to our blog on using great expectations with Dagster.

Or if we consider the GE integration complete, referencing the integration and that testing frameworks can be leveraged to ease this process.

Author: Done in [dc082fc](https://github.com/dagster-io/dagster/pull/24035/commits/dc082fc8c605309d6c9b88d058f000cbc71d465a)

</details>

## Testing assets with a single asset check

The example below defines a single asset check on an asset that fails if the `order_id` column of the asset contains a null value.

In this example, the asset check will run after the asset has been materialized, to ensure the quality of its data.

<CodeExample filePath="guides/data-assets/quality-testing/asset-checks/single-asset-check.py" language="python" title="Asset with a single asset check" />

## Testing assets with multiple asset checks

In most cases, checking the data quality of an asset will require multiple checks.

The example below defines two asset checks using the `multi_asset_check` decorator:
- one that fails if the `order_id` column of the asset contains a null value.
- another one that fails if the `item_id` column of the asset contains a null value.

In this example, both asset check will run in a single operation after the asset has been materialized.

<CodeExample filePath="guides/data-assets/quality-testing/asset-checks/multiple-asset-checks.py" language="python" title="Asset with multiple asset checks" />

## Defining asset checks using a factory pattern

> Defining multiple checks can also be done using a factory pattern. The example below defines the same two asset checks as in the previous example, but this time using a factory pattern and the `@asset_check` decorator.

Reviewer: Curious why you chose this over a `multi_asset_check`. Wouldn't this run `n` _ops_, one for each check?

Author: That's a good call, I've updated the example to use multi_asset_check [here](https://github.com/dagster-io/dagster/pull/24035/commits/5c38554cca8501735ad628bb6157a7cd58bbb583)

<summary>Prerequisites</summary>
- Familiarity with [Assets](/concepts/assets)
</details>

## Testing assets with a single asset check

The example below defines a single asset check on an asset that fails if the `order_id` column of the asset contains a null value.

In this example, the asset check will run after the asset has been materialized, to ensure the quality of its data.

<CodeExample filePath="guides/data-assets/quality-testing/asset-checks/single-asset-check.py" language="python" title="Asset with a single asset check" />

## Testing assets with multiple asset checks

In most cases, checking the data quality of an asset will require multiple checks.

The example below defines two asset checks using the `@multi_asset_check` decorator:

- One check that fails if the `order_id` column of the asset contains a null value
- Another check that fails if the `item_id` column of the asset contains a null value

In this example, both asset checks will run in a single operation after the asset has been materialized.

<CodeExample filePath="guides/data-assets/quality-testing/asset-checks/multiple-asset-checks.py" language="python" title="Asset with multiple asset checks" />

> ## Defining asset checks using a factory pattern

Reviewer 2: I'd title this section "Programmatically generating asset checks" or something along those lines. It's a bit of a nitpick, but in my mind the factory pattern is just a tool to achieve the goal of "generate lots of similar asset checks", and so someone with that goal in mind would be better served by a title oriented around that.

> ## Defining asset checks using a factory pattern

Reviewer: ```suggestion
## Programmatically generating asset checks
```

Defining multiple checks can also be done using a factory pattern. The example below defines the same two asset checks as in the previous example, but this time using a factory pattern and the `@multi_asset_check` decorator.

<CodeExample filePath="guides/data-assets/quality-testing/asset-checks/asset-checks-factory.py" language="python" title="Defining asset checks using a factory pattern" />

## Blocking downstream assets

> By default, materialization of downstream assets will continue even if a parent's asset check fails. To block the materialization of downstream assets, set the `blocking` argument to `True` in the `@asset_check` decorator.

Reviewer: ```suggestion
By default, if a parent's asset check fails during a run, the run will continue and downstream assets will be materialized. To prevent this behavior, set the `blocking` argument to `True` in the `@asset_check` decorator. This will cause the run to fail and prevent further materializations.
```

> By default, materialization of downstream assets will continue even if a parent's asset check fails. To block the materialization of downstream assets, set the `blocking` argument to `True` in the `@asset_check` decorator.

Reviewer 2: this section should emphasize that this blocking behavior is within the context of a run. You're still able to manually materialize the downstream asset after a blocking check fails. Maybe:

> By default, if a parent's asset check fails during a run, the run will continue and downstream assets will be materialized. To prevent this behavior, set the `blocking` argument to `True` in the `@asset_check` decorator. This will cause the run to fail and prevent further materializations.
